gapis/api/vulkan/transform_query_timestamps.go

```diff
@@ -0,0 +1,624 @@
+	beginCommandBufferInfo := NewVkCommandBufferBeginInfo(inputState.Arena,
+		VkStructureType_VK_STRUCTURE_TYPE_COMMAND_BUFFER_BEGIN_INFO, // sType
+		0, // pNext
+		VkCommandBufferUsageFlags(VkCommandBufferUsageFlagBits_VK_COMMAND_BUFFER_USAGE_ONE_TIME_SUBMIT_BIT), // flags
+		0, // pInheritanceInfo
+	)
+	beginCommandBufferInfoData := timestampTransform.allocations.AllocDataOrPanic(ctx, beginCommandBufferInfo)
+
+	outputCmds := make([]api.Cmd, 0)
+
+	outputCmds = append(outputCmds,
+		cb.VkAllocateCommandBuffers(
+			device,
+			commandBufferAllocateInfoData.Ptr(),
+			commandBufferData.Ptr(),
+			VkResult_VK_SUCCESS,
+		).AddRead(
+			commandBufferAllocateInfoData.Data(),
+		).AddWrite(
+			commandBufferData.Data(),
+		),
+		cb.VkBeginCommandBuffer(
+			commandBufferID,
+			beginCommandBufferInfoData.Ptr(),
+			VkResult_VK_SUCCESS,
+		).AddRead(
+			beginCommandBufferInfoData.Data(),
+		),
+		cb.VkCmdResetQueryPool(commandBufferID, info.queryPool, info.queryCount, 1),
+		cb.VkCmdWriteTimestamp(commandBufferID,
+			VkPipelineStageFlagBits_VK_PIPELINE_STAGE_BOTTOM_OF_PIPE_BIT,
+			info.queryPool,
+			info.queryCount),
+		cb.VkEndCommandBuffer(
+			commandBufferID,
+			VkResult_VK_SUCCESS,
+		))
+
+	return outputCmds, commandBufferID
+}
+
+func (timestampTransform *queryTimestamps) processNotification(ctx context.Context, s *api.GlobalState, n gapir.Notification) {
+	notificationData := n.GetData()
+	notificationID := n.GetId()
+	timestampsData := notificationData.GetData()
+
+	res, ok := timestampTransform.results[notificationID]
+	if !ok {
+		log.I(ctx, "Invalid notificationID %d", notificationID)
+		return
+	}
+
+	byteOrder := s.MemoryLayout.GetEndian()
+	r := endian.Reader(bytes.NewReader(timestampsData), byteOrder)
+	tStart := r.Uint64()
+	var timestamps service.Timestamps
+	resultCount := uint32(len(timestampsData) / 8)
+	resIdx := 0
+
+	for i := uint32(1); i < resultCount; i++ {
+		tEnd := r.Uint64()
+		record := res[resIdx]
+		record.timestamp.TimeInNanoseconds = uint64(float32(tEnd-tStart) * timestampTransform.timestampPeriod)
+		if record.IsEoC {
+			tStart = r.Uint64()
+			i++
+		} else {
+			tStart = tEnd
+		}
+		timestamps.Timestamps = append(timestamps.Timestamps, &record.timestamp)
+		resIdx++
+	}
+
+	timestampTransform.handler(&service.GetTimestampsResponse{
+		Res: &service.GetTimestampsResponse_Timestamps{
+			Timestamps: &timestamps,
+		},
+	})
+}
+
+func (timestampTransform *queryTimestamps) cleanup(ctx context.Context, inputState *api.GlobalState) []api.Cmd {
+	cb := CommandBuilder{Thread: 0, Arena: inputState.Arena}
+	outputCmds := make([]api.Cmd, 0, len(timestampTransform.queryPools)+len(timestampTransform.commandPools))
+
+	for _, info := range timestampTransform.queryPools {
+		cmd := cb.VkDestroyQueryPool(
+			info.device,
+			info.queryPool,
+			memory.Nullptr,
+		)
+		outputCmds = append(outputCmds, cmd)
+	}
+
+	// Melih TODO: ask about this
+	// for commandPoolkey, commandPool := range timestampTransform.commandPools {
```

Comment:
???

Reply:
PM

Reply:
Has this been clarified?

Also, I think it would be better to have a discussion here (rather than on private messages) for all reviewers to benefit from it :)